compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -239,72 +296,149 @@ void ReductionStrategy3StageThreadDistribution::compute(
+ public:
+  /// `hasTrailingElementwise` is currently used to guard against pathological
+  /// cases where IREE can't bound a buffer and crashes.
+  // TODO: Fix IREE's odegen/Common/PadDynamicAlloc.cpp.
+  SmallReductionStrategy(
+      MLIRContext *context,
+      const transform_ext::MatchedReductionCaptures &captures,
+      int64_t maxNumThreadsToUse, bool hasTrailingElementwise)
+      : GPUReductionStrategy(context, captures) {
+    compute(maxNumThreadsToUse, hasTrailingElementwise);
   }
 
-  void computeThreadDistribution(int64_t maxNumThreads,
-                                 int64_t warpShuffleSize) {
-    threadDistribution3Stages =
-        std::make_unique<ReductionStrategy3StageThreadDistribution>(
-            reductionDimensionSize, maxNumThreads, warpShuffleSize);
+  SmallReductionStrategy(const SmallReductionStrategy &) = default;
+  SmallReductionStrategy &operator=(const SmallReductionStrategy &) = default;
+
+  int64_t getNumThreadsXInBlock() const { return getNumThreadsInBlock()[0]; }
+  int64_t getNumThreadsYInBlock() const { return getNumThreadsInBlock()[1]; }
+  int64_t getNumThreadsZInBlock() const { return getNumThreadsInBlock()[2]; }
+  std::array<int64_t, 3> getNumThreadsInBlock() const override {
+    std::array<int64_t, 3> res{1, 1, 1};
+    for (int64_t i = 0, e = workgroupTileSizes.size(); i < e; ++i)
+      res[i] = workgroupTileSizes[i];
+    return res;
   }
 
-  /// Constructor quantities.
-  MLIRContext *context;
-  int64_t reductionRank;
-  int64_t reductionDimensionSize;
-  int64_t maybeLeadingRank;
-  int64_t maybeTrailingRank;
+  bool isValid() override { return valid; }
 
-  /// Derived quantities.
-  std::unique_ptr<ReductionStrategy3StageThreadDistribution>
-      threadDistribution3Stages;
-  SmallVector<Attribute> allBlockAttrs;
-  // Tile sizes for the workgroup / determines grid size.
-  SmallVector<int64_t> workgroupTileSizes;
-  // Launch bounds for the workgroups / block size.
-  std::array<int64_t, 3> workgroupSize;
+ private:
+  bool valid = false;
+
+  /// Compute the small strategy based on the problem size and the
+  /// `maxNumThreadsToUse`.
+  /// `hasTrailingElementwise` is currently used to guard against pathological
+  /// cases where IREE can't bound a buffer and crashes.
+  // TODO: Fix IREE's odegen/Common/PadDynamicAlloc.cpp.
+  void compute(int64_t maxNumThreadsToUse, bool hasTrailingElementwise);
 };
+
+void SmallReductionStrategy::compute(int64_t maxNumThreadsToUse,
+                                     bool hasTrailingElementwise) {
+  assert(maxNumThreadsToUse > 0 && "maxNumThreadsToUse must be > 0");
+  assert(maxNumThreadsToUse >= iree_compiler::kCudaWarpSize &&
+         "not even a warp?");
+
+  // Block-level
+  // ===========
+  FailureOr<int64_t> maybeDivisor =
+      mlir::iree_compiler::maxDivisorOfValueBelowLimit(
+          captures.mostMinorParallelDimensionSize, maxNumThreadsToUse);
+
+  // Dynamic reductions are never supported by default.
```

Comment:
... because we never know we are in a small-reduction regime mode.
Since this mode does not coalesce reads, perf will often suffer badly.